src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/SNI/SNITcpHandle.cs

```diff
@@ -322,112 +324,76 @@ private Socket TryConnectParallel(string hostName, int port, TimeSpan ts, bool i
         // Only write to the DNS cache when we receive IsSupported flag as true in the Feature Ext Ack from server.
         private static Socket Connect(string serverName, int port, TimeSpan timeout, bool isInfiniteTimeout, string cachedFQDN, ref SQLDNSInfo pendingDNSInfo)
         {
-            IPAddress[] ipAddresses = Dns.GetHostAddresses(serverName);
+            // keeping IEnumerable to enumerate entire array only if needed
+            IEnumerable<IPAddress> ipAddresses = Dns.GetHostAddresses(serverName);
+            
+            // First, we try to connect all AddressFamily.InterNetwork
+            // and then all AddressFamily.InterNetworkV6.
+            // Keeping address families inlined in entire method because parameter pendingDNSInfo is bound on them.
+            ipAddresses = ipAddresses
+                .Where (address => address.AddressFamily==AddressFamily.InterNetwork)
+                .Concat(ipAddresses.Where(address => address.AddressFamily==AddressFamily.InterNetworkV6));
```

Comment:
Can't this be achieved with an `OrderBy`? Also if this is on a hot path maybe linq should be avoided

Reply:
It can be achieved with OrderBy, I made so firstly: https://github.com/dotnet/SqlClient/pull/1029/commits/0021191c41c0ce8e5f5b091426961b7351aa05fe
But, that is more code with more enumeration. While this one is shorter and does not enumerate everything upon needed.
This is not hot path, we use exception/async here as well and even thrown exception does not play any difference here.

Reply:
Here some information about performance https://github.com/dotnet/SqlClient/issues/826#issuecomment-739660666

Reply:
@DavoudEshtehari Helpful for what?
I have updated this PR with https://github.com/dotnet/SqlClient/pull/1016 already.
I don't see anything else in 1015 , which could be in contrast with this PR. That code had same issues which already discussed in 1016

Reply:
I understand you're trying to improve the socket establishment here, and our focus on PR [#1015](https://github.com/dotnet/SqlClient/pull/1015) is to add a new configuration not performance improvement. The new IP preference feature uses the resolved IPs from dns in original order that is important especially for the **default preference**. This is a friendly point should be considered here if there is any chance of reordering the IP addresses. For more explanation you can check [this comment](https://github.com/dotnet/SqlClient/pull/1015#issuecomment-812609864).

Reply:
@DavoudEshtehari  I see. Well, this PR is not bound on socket selecting method. The point of this PR is to fix issues listed under #826 , for example #449 , or for instance to fix timeout issue described in #1016 . Any IP addresses ordering is fine for this PR.
What I'm saying, is I've already changed the addresses order once to match #1016 and it can be updated again to do any ordering.
However, as I see it, it's easier to merge this PR with any ordering we want and then to update other conflicting PRs, because this one makes code 20% shorter:
![image](https://user-images.githubusercontent.com/10516222/118206671-4954fa80-b463-11eb-86b0-5609399ed4e5.png)
Anyway, I think, it does not make sense to update this PR each time we change ordering in main branch. Probably it would be more efficient to update this PR once right before the merge.

Reply:
I think it would be a good time to prepare your final change if you had time to back to this PR.